driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -0,0 +1,75 @@
+use super::*;
+use crate::contracts::stablex_contract::batch_exchange;
+use anyhow::{anyhow, bail, Result};
+use block_timestamp::BlockTimestamp;
+use ethcontract::{contract::Event, errors::ExecutionError};
+use futures::stream::{BoxStream, StreamExt as _};
+use orderbook::Orderbook;
+use std::future::Future;
+use std::sync::{mpsc, Arc, RwLock};
+
+// TODO:
+// `pub struct UpdatingOrderbook` that also owns an `Arc<RwLock<Orderbook>>` and creates a
+// background thread running `UpdatingOrderbookThread::update_with_events`.
+// This way `UpdatingOrderbookThread` is independently testable.
+
+/// Update the orderbook with events from the stream forever or until exit_indicator is dropped.
+///
+/// Returns Ok when exit_indicator is dropped.
+/// Returns Err if the stream ends.
+async fn update_with_events_forever(
+    orderbook: Arc<RwLock<Orderbook>>,
+    mut block_timestamp: impl BlockTimestamp,
+    exit_indicator: mpsc::Receiver<()>,
+    past_events: impl Future<Output = Result<Vec<Event<batch_exchange::Event>>, ExecutionError>>,
+    mut stream: BoxStream<'static, Result<Event<batch_exchange::Event>, ExecutionError>>,
+) -> Result<()> {
+    let should_exit = || {
+        matches!(
+            exit_indicator.try_recv(),
+            Err(mpsc::TryRecvError::Disconnected)
+        )
+    };
+
+    for event in past_events.await? {
+        handle_event(&orderbook, &mut block_timestamp, event).await?;
+    }
```

Comment:
I think you should lock the `RwLock` once instead of once per iteration here.

Reply:
I was considering this but then if someone tries to get the orderbook it would block until we have handled all past events. On the other hand  you wouldn't get a useful orderbook anyway if we are still handling past events.

Reply:
Ahhhh, good point. I forgot that the block timestamp needs to be queried in the `handle_event`, yes then the lock should definitely happen once per event.